clang/lib/3C/CastPlacement.cpp

```diff
@@ -172,33 +172,31 @@ CastPlacementVisitor::getCastString(ConstraintVariable *Dst,
   case CAST_TO_WILD:
     return std::make_pair("((" + Dst->getRewritableOriginalTy() + ")", ")");
   case CAST_TO_CHECKED: {
-    std::string Suffix = ")";
+    std::string Type, Suffix;
     if (const auto *DstPVC = dyn_cast<PVConstraint>(Dst)) {
       assert("Checked cast not to a pointer" && !DstPVC->getCvars().empty());
       ConstAtom *CA =
           Info.getConstraints().getAssignment(DstPVC->getCvars().at(0));
 
-      // Writing an _Assume_bounds_cast to an array type requires inserting
-      // the bounds for destination array. These can come from the source
-      // code or the infered bounds. If neither source is available, use empty
-      // bounds.
-      if (isa<ArrAtom>(CA) || isa<NTArrAtom>(CA)) {
-        std::string Bounds = "";
-        if (DstPVC->srcHasBounds())
-          Bounds = DstPVC->getBoundsStr();
-        else if (DstPVC->hasBoundsKey())
-          Bounds = ABRewriter.getBoundsString(DstPVC, nullptr, true);
-        if (Bounds.empty())
-          Bounds = "byte_count(0)";
-
-        Suffix = ", " + Bounds + ")";
+      // TODO: Writing an _Assume_bounds_cast to an array type requires
+      // inserting the bounds for destination array. But the names used in src
+      // and dest may be different, so we need more sophisticated code to
+      // convert to local variable names. Use unknown bounds for now.
+      if (isa<ArrAtom>(CA)) {
+        Type = "_Array_ptr<";
```

Comment:
I think I would prefer only setting `Type` to the output pointer type in the `TypeVar != nullptr` branch where it's actually used.

It also seems to me that this duplicates a little bit of the logic in `mkString`. That code is already responsible for looking at `ConstAtom`s and emitting the correct pointer type string. I recall that there was a reason you couldn't just call `TypeVar->mkString` with `EmitPointee = false`, but that still feels the right thing to do if whatever problem was stopping it can reasonably be avoided.

Reply:
I think the issue here is that there is no target to run the `mkString` on. If the `TypeVar` exists it means the local typevar is different from the destination typevar, so we can't use the destination type. But the local complete type is being cast to the destination type, so we can't use the local one either either. 

If I'm missing some set of parameters to `mkString` that solves this, let me know. I don't know all the combinations.

Reply:
I just tried to set `Type` later as you suggested and it would involve duplicating all the surrounding code. Because it is `Suffix` that needs to be done there, with access to the destination CA, and `Type` is set out of convenience.

Reply:
I spent some more time looking at this, and, while I'm still not entirely happy with how this seems to re-implement a small part of the mkstring logic, I was not able to find a implementation I like better.